### PR TITLE
borg benchmark crud command

### LINKED ISSUE #1788: borg benchmark command
It would be nice to have a command to benchmark different subsystems.   Something like

```
borg benchmark repo /mnt/external_drive/backup
     100MB/s up   200MB/s down
borg benchmark repo offsetitebackup.com:foo
     1MB/s up   10MB/s down
borg benchmark compression lzma,1
      2MB/s compression  8MB/s decompression
borg benchmark chunker
   ... etc
```

My current use case is picking compression settings for on-site and off-site backups, as a new user.  Today answering a question "how much compression can I run and still be IO bound" is fairly involved, takes time and knowledge to setup and run tests.   With builtin benchmarking support it could be done by a simple script in a few minutes.

Another usecase would be trouble-shooting performance issues by users/sysadmins (it is network, hard drive, code regression?).  Also, it would be useful for development, to trend performance over time.

---

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/38830536-borg-benchmark-command)

(not necessarily like described above, see also comments)

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
fixes #1788.

result looks like this:
```
$ BORG_PASSPHRASE=secret borg init --encryption=repokey-blake2 repo
$ BORG_PASSPHRASE=secret borg benchmark crud repo testdir
C-Z-BIG         122.27 MB/s (10 * 100.00 MB all-zero files: 8.18s)
R-Z-BIG         190.21 MB/s (10 * 100.00 MB all-zero files: 5.26s)
U-Z-BIG         516.33 MB/s (10 * 100.00 MB all-zero files: 1.94s)
D-Z-BIG         850.42 MB/s (10 * 100.00 MB all-zero files: 1.18s)
C-R-BIG          43.38 MB/s (10 * 100.00 MB random files: 23.05s)
R-R-BIG         133.22 MB/s (10 * 100.00 MB random files: 7.51s)
U-R-BIG         534.41 MB/s (10 * 100.00 MB random files: 1.87s)
D-R-BIG         231.92 MB/s (10 * 100.00 MB random files: 4.31s)
C-Z-MEDIUM       93.89 MB/s (1000 * 1.00 MB all-zero files: 10.65s)
R-Z-MEDIUM      209.07 MB/s (1000 * 1.00 MB all-zero files: 4.78s)
U-Z-MEDIUM      720.55 MB/s (1000 * 1.00 MB all-zero files: 1.39s)
D-Z-MEDIUM      857.18 MB/s (1000 * 1.00 MB all-zero files: 1.17s)
C-R-MEDIUM       46.75 MB/s (1000 * 1.00 MB random files: 21.39s)
R-R-MEDIUM      128.95 MB/s (1000 * 1.00 MB random files: 7.75s)
U-R-MEDIUM      699.98 MB/s (1000 * 1.00 MB random files: 1.43s)
D-R-MEDIUM      245.63 MB/s (1000 * 1.00 MB random files: 4.07s)
C-Z-SMALL        19.29 MB/s (10000 * 10.00 kB all-zero files: 5.18s)
R-Z-SMALL        96.73 MB/s (10000 * 10.00 kB all-zero files: 1.03s)
U-Z-SMALL        40.41 MB/s (10000 * 10.00 kB all-zero files: 2.47s)
D-Z-SMALL        79.66 MB/s (10000 * 10.00 kB all-zero files: 1.26s)
C-R-SMALL        10.97 MB/s (10000 * 10.00 kB random files: 9.11s)
R-R-SMALL        91.39 MB/s (10000 * 10.00 kB random files: 1.09s)
U-R-SMALL        67.10 MB/s (10000 * 10.00 kB random files: 1.49s)
D-R-SMALL        57.06 MB/s (10000 * 10.00 kB random files: 1.75s)
```
Notes:
- C-Z- is borg create with all-zero files. due to dedup and no files cache, this is primarily testing reader + chunker + hasher speed.
- C-R- is borg create with pure random files. no dedup, this is measuring throughput through all processing stages.
- R- is borg extract --dry-run (doing everything except writing result to disk)
- U- is testing borg create 2nd run with unchanged input dataset, files cache speed. the throughput value is kind of virtual here, it does not actually read the file.
- D- is testing borg delete, removing the LAST archive that references the data.
- there is quite some variance in these measurements, good measurements need to be longer (and on a otherwise idle machine)

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
codecov is ofc whining about the missing test, but due to the rather long runtime, i guess it is better to not include this in the tests (at least not "as is").

i am considering adding some scale factor for count, so that instead of testing with ~1GB one can also have it test with 0.1 or 10 or 100 GB. Maybe 0.1 is still too much for the tests, though, the small files test preparation takes quite long for generating these.

File: src/borg/archiver.py [hunk 1]
```diff
@@ -315,6 +317,72 @@ def do_migrate_to_repokey(self, args, repository):
         logger.info('Key updated')
         return EXIT_SUCCESS
 
+    def do_benchmark_crud(self, args):
+        def measurement_run(repo, path):
+            archive = repo + '::borg-benchmark-crud'
+            compression = '--compression=none'
+            # measure create perf (without files cache to always have it chunking)
+            t_start = time.time()
```
Comment (enkore) [inline src/borg/archiver.py]: .monotonic everywhere?

Comment (ThomasWaldmann) [inline src/borg/archiver.py]: yup

**Comment by ThomasWaldmann** [comment]
anything else?

**Comment by ThomasWaldmann** [comment]
OK, then I'll collapse the changesets and add some docs (like stuff given in top post).

**Comment by enkore** [comment]
Two feature additions come to mind

1. Scale factor / --how-much-data like you mentioned previously
2. Using a faster RNG (we have the AES() class, we could just encrypt 32K of zeroes while counting IV, guaranteed to be faster than the rather slow /dev/urandom).

File: src/borg/archiver.py [hunk 2]
```diff
@@ -3141,6 +3209,59 @@ def process_epilog(epilog):
...
+                                          description='benchmark command',
+                                          epilog=benchmark_epilog,
+                                          formatter_class=argparse.RawDescriptionHelpFormatter,
+                                          help='benchmark command')
+
+        benchmark_parsers = subparser.add_subparsers(title='required arguments', metavar='<command>')
+        subparser.set_defaults(fallback_func=functools.partial(self.do_subcommand_help, subparser))
+
+        bench_crud_epilog = process_epilog("""
+        This command benchmarks borg CRUD (create, read, update, delete) operations.
```
Comment (enkore) [inline src/borg/archiver.py]: While the command line does say this, also repeat here, that this works on a real repository and needs real free space there and does real modifications on it?

Comment (ThomasWaldmann) [inline src/borg/archiver.py]: yup, added some clarifications.

**Comment by ThomasWaldmann** [comment]
about using AES: i tested the cmd on my laptop, urandom() is fast enough. so I'll rather keep it simple.
in case we need more pseudo-random data, one could also "spray" counters over random data to reuse it multiple times. spraying into locations tighter together than chunker granularity.

scale factor: yes, separate PR.

**Comment by enkore** [comment]
Oh, seems urandom was updated since I last looked at it. I remembed double-digit speeds (like 30 MB/s), but now it gives ~260 MB/s which is quite ok.

**Pull request merged**